### ISSUE TITLE
Spring Method Security does not seem to work correctly.

### ISSUE THREAD

**Issue by Nick-Wunderdog** [paulcwarren/spring-content#1374]
**Describe the bug**
Spring Method level Security is not working as expected. It seems that the class level `@PreAuthorize("false") ` overrides the method level `@PreAuthorize("true") ` in classes extending `ContentStore<Attachment, String>` in Spring the 

**To Reproduce**
Create ContentStore class: 

```java
@StoreRestResource
@PreAuthorize("false")
public interface AttachmentContentStore extends ContentStore<Attachment, String> {
    String WRITE_PERMISSION = "true";
    String READ_PERMISSION = "true";

    @Override
    @PreAuthorize(WRITE_PERMISSION)
    Attachment setContent(Attachment attachment, InputStream inputStream);

    @Override
    @PreAuthorize(WRITE_PERMISSION)
    Attachment setContent(Attachment attachment, PropertyPath propertyPath, InputStream inputStream);

    @Override
    @PreAuthorize(WRITE_PERMISSION)
    Attachment setContent(Attachment attachment, PropertyPath propertyPath, InputStream inputStream, long l);

    @Override
    @PreAuthorize(WRITE_PERMISSION)
    Attachment setContent(Attachment attachment, Resource resource);

    @Override
    @PreAuthorize(WRITE_PERMISSION)
    Attachment setContent(Attachment attachment, PropertyPath propertyPath, Resource resource);

    @Override
    @PreAuthorize(WRITE_PERMISSION)
    void associate(Attachment attachment, String s);

    @Override
    @PreAuthorize(WRITE_PERMISSION)
    void associate(Attachment attachment, PropertyPath propertyPath, String s);

    @Override
    @PreAuthorize(WRITE_PERMISSION)
    Attachment unsetContent(Attachment attachment);

    @Override
    @PreAuthorize(WRITE_PERMISSION)
    Attachment unsetContent(Attachment attachment, PropertyPath propertyPath);

    @Override
    @PreAuthorize(WRITE_PERMISSION)
    void unassociate(Attachment attachment);

    @Override
    @PreAuthorize(WRITE_PERMISSION)
    void unassociate(Attachment attachment, PropertyPath propertyPath);



    @Override
    @PreAuthorize(READ_PERMISSION)
    InputStream getContent(Attachment attachment);

    @Override
    @PreAuthorize(READ_PERMISSION)
    InputStream getContent(Attachment attachment, PropertyPath propertyPath);

    @Override
    @PreAuthorize(READ_PERMISSION)
    Resource getResource(Attachment attachment);

    @Override
    @PreAuthorize(READ_PERMISSION)
    Resource getResource(Attachment attachment, PropertyPath propertyPath);

    /** Called by get content HTTP client */
    @Override
    @PreAuthorize(READ_PERMISSION)
    Resource getResource(Attachment attachment, PropertyPath propertyPath, GetResourceParams getResourceParams);

    @Override
    @PreAuthorize(READ_PERMISSION)
    Resource getResource(String s);

}
```
**Expected behavior**
Both PUT /attachments/1/content and GET /attachments/1/content should succeed with HTTP 200 (or 201 if created) 
But both with above code the both fail with HTTP 403. Which would be correct response if **methods**  were annotated with 
`@PreAuthorize("false")`
 
Code works fine without Spring security annotaitons and some combinations of annotation work, but that is not the desired combination. The expected behavior is that the method annotation should overide the class level annotation. This does not work, also methods annotated with READ_PERMISSION do not seem to work at all. 

The result matrix is on The Spring Security PreAuthorize value variations: 
(PostAuthorize gives identical results)
| Req  | class | WRITE | READ | resp |                  |
|--------|---------|-----------|---------|--------|---------------|
|  PUT | false | true    | true    |  403 | incorrect  |
|  GET | false | true    | true    | 403  | incorrect  |
|  PUT | true | false   | false    |  403 | correct     |
|  GET | true | false   | false    | 200  | incorrect  |
|  PUT |         | false   | false    |  403 | correct     |
|  GET |         | false   | false    | 200  | incorrect (x) |
|  PUT |         | true    | true    |  200 | correct     |
|  GET |         | true    | true    | 200  | correct (x) |

x = method level annotation do not seem to have any effect on these
Rows with empty "class" column mean the class level annotation was omitted. 

**Screenshots**
N/A

**Additional context**
using :

curl -X PUT -H 'Content-Type:text/plain' -d 'Hello Spring Content World!' http://localhost:8080/attachments/1/content
curl -X GET -H 'Content-Type:text/plain'  http://localhost:8080/atttachments/1/content

**Comment by Nick-Wunderdog** [comment]
And let me say: Thanks for great project and all the hard work on it.

**Comment by paulcwarren** [comment]
I see.  Thanks for the report @Nick-Wunderdog.  Appreciate the detailed report.  

The project is definitely missing these test cases so I am not surprised there are inconsistencies here.  Will look into soonest.

**Comment by Nick-Wunderdog** [comment]
Thank  you once again. It is very stange as it kind of works, but not correctly. 
1. Some of the read methods do not seem to abide Spring (Method) Security at all
2. Presedence of class level and method level annotations seem to be wrong. Annotation on method should override the class level annotation.

**Comment by paulcwarren** [comment]
I've debugged this a little bit today and I cant reproduce some of your findings thus far.  Moreover, Repository and ContentStore behavior appear identical to me.  

I'm playing around with a repository as follows:

```
@PreAuthorize("false")
public interface FileRepository extends JpaRepository<File, Long> {

    @Override
    @PreAuthorize("true")
    public Optional<File> findById(Long id);

    @Override
    @PreAuthorize("true")
    public File save(File entity);
}
```

And a ContentStore as follows:

```
@PreAuthorize("false")
public interface FileContentStore extends ContentStore<File, String> {

	@PreAuthorize("true")
	@Override
	Resource getResource(File entity, PropertyPath propertyPath, GetResourceParams params);

	@PreAuthorize("true")
	@Override
	File setContent(File entity, PropertyPath propertyPath, InputStream content, long contentLen);
}
```

And, for example, when I POST, or PUT to `/files/` and `/files/{id}`.  Or GET from `/files/{id}` it is also allowed.  i.e. to the Spring Data repository.  Likewise when I PUT or GET to `/files/{id}/content`; i.e. the content store, it is also allowed.   In all cases the `AuthorizationManagerBeforeMethodInterceptor` is used to perform the auth checks.

If I flip the PreAuthorize's (class == true, findById & getResource == false) then GET `/files/{id}` and GET `/files/{id}/content` then both return 403.

I'll continue to try and reproduce your findings but one point to note is that Spring Data REST and Spring Content REST layers both perform `findById` calls to find the entity that is being acted on.  In addition SCR performs a `getResource` call and then potentially a `setContent` and a repository `save`.  I noticed in my own testing it is easy to accidentally lock operations down.  Just a thought.

**Comment by Nick-Wunderdog** [comment]
Strange. 
I am using Temurin 17 Java, latest build + Spring Boot 3.0.5 parent


```bash
curl -X PUT "{host}/api/attachments/{id}/content" -H "Content-Type: text/plain" -d "Hello"
```

Debugging that call in `ObservationAuthorizationManager.check()` method the authorization flow  is : 
1. RequestMatcherDelegatingAuthrizationManager.check() = true
2. RequestMatcherDelegatingAuthrizationManager.check() = true
3. PreAuthorizeAuthrizationManager.check() = true                          // from JPA repo findById() METHOD PreAuth 
4. PostAuthorizeAuthrizationManager.check() = true                        // from JPA repo findById() METHOD PostAuth (no class level PostAuth) 
5. PreAuthorizeAuthrizationManager.check() = **false** // from Store CLASS PreAuth!!!!

So the difference for me is that while identically annotated JPA repoclass method level annotation overrrides the class level annotation and the class level `@PerAuthorize` in evaluated. In ContentStore it always seem to evaluate the Class level annotation. 

If I remove the Class level annotation from ContentStore then it works and checks the method level annnotation the PUT works. This could be somewhat acceptable solution, albeit not secure. **But** now without class level security annotation the method level security annotation is never called for GET 

curl -X GET "{host}/api/attachments/{id}/content"  -H "Content-Type: text/plain" \

 It does return the content, but has no access control at all.

**Comment by Nick-Wunderdog** [comment]
@paulcwarren I am trying to work around this problem. Can  you confirm that there are no direct HTTP access to the content ? 
That there is no "/content/{content_id} mapping or something to access the content without internally first retrieving the JPA parent entity  "/files/{jpa_entity_id}/content" ? 

That would defacto secure the content access via the parent entity JPA repository security.

**Comment by paulcwarren** [comment]
Still debugging.  I added your debug service to my appp and ran a test that gets the entity, gets entity content and set entity content and I get this output.  Note, its is sometimes different:

```
# these calls are direct to the JAVA API
****** DEBUG: {FileRepository save security} ******
****** DEBUG: {FileContentStore setContent security} ******
****** DEBUG: {FileRepository save security} ******

# these calls a via SDR/SCR
****** DEBUG: {FileContentStore class level security} ******
****** DEBUG: {FileRepository class level security} ******
****** DEBUG: {FileRepository class level security} ******
****** DEBUG: {FileContentStore class level security} ******
****** DEBUG: {FileRepository findById security} ******
****** DEBUG: {FileContentStore class level security} ******
****** DEBUG: {FileRepository class level security} ******
****** DEBUG: {FileRepository class level security} ******
****** DEBUG: {FileContentStore class level security} ******
****** DEBUG: {FileRepository findById security} ******
****** DEBUG: {FileContentStore getResource security} ******
****** DEBUG: {FileContentStore class level security} ******
****** DEBUG: {FileRepository class level security} ******
****** DEBUG: {FileRepository class level security} ******
****** DEBUG: {FileContentStore class level security} ******
****** DEBUG: {FileRepository findById security} ******
****** DEBUG: {FileContentStore getResource security} ******
****** DEBUG: {FileContentStore class level security} ******
****** DEBUG: {FileRepository save security} ******
```

Anyway, no SCR does not offer any direct to content API for `ContentStore`s.

**Comment by Nick-Wunderdog** [comment]
I used IDE(A) automation to override all the methods form ContentStore and I added `@PreAuthorize` on all of them. 
I rewritten the tests several time and I consistently get same result that the method level `@PreAuthorize("true")` is never called in ContentStore class if there is class level `@PreAuthorize("false")` or vice versa. There is something really strange that we are getting differnet results. 
I sent you an email with my contact details, in case you want to have a call and see this live. 

Here is my pom.xml 

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>3.0.5</version>
        <relativePath/> <!-- lookup parent from repository -->
    </parent>
    <groupId>com.example</groupId>
    <artifactId>example</artifactId>
    <version>0.0.1-SNAPSHOT</version>
    <name>Example</name>
    <description>Example to reproduce Spring security problem</description>
    <properties>
        <java.version>17</java.version>
        <file.encoding>UTF-8</file.encoding>
        <skip.npm.build>false</skip.npm.build>
    </properties>
    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-starter</artifactId>
                <exclusions>
                    <exclusion>
                        <groupId>org.springframework.boot</groupId>
                        <artifactId>spring-boot-starter-logging</artifactId>
                    </exclusion>
                </exclusions>
            </dependency>
            <dependency>
                <groupId>org.springframework.data</groupId>
                <artifactId>spring-data-bom</artifactId>
                <version>2022.0.2</version>
                <scope>import</scope>
                <type>pom</type>
            </dependency>
            <dependency>
               <groupId>com.github.paulcwarren</groupId>
                <artifactId>spring-content-jpa</artifactId>
                <exclusions>
                    <exclusion>
                        <groupId>commons-logging</groupId>
                        <artifactId>commons-logging</artifactId>
                    </exclusion>
                </exclusions>
            </dependency>
        </dependencies>
    </dependencyManagement>
    <dependencies>

        <!-- Spring boot -->
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-actuator</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-data-jpa</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.data</groupId>
            <artifactId>spring-data-jpa</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.data</groupId>
            <artifactId>spring-data-envers</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-data-rest</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-hateoas</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-web</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-security</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-oauth2-resource-server</artifactId>
        </dependency>

        <!-- DB -->
        <dependency>
            <groupId>com.h2database</groupId>
            <artifactId>h2</artifactId>
        </dependency>
        <dependency>
            <groupId>org.hibernate.validator</groupId>
            <artifactId>hibernate-validator</artifactId>
            <version>8.0.0.Final</version>
        </dependency>
        <dependency>
            <groupId>org.flywaydb</groupId>
            <artifactId>flyway-core</artifactId>
        </dependency>
        <dependency>
            <groupId>com.github.paulcwarren</groupId>
            <artifactId>spring-content-jpa-boot-starter</artifactId>
            <version>3.0.1</version>
            <exclusions>
                <exclusion>
                    <groupId>commons-logging</groupId>
                    <artifactId>commons-logging</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>jakarta.servlet</groupId>
                    <artifactId>jakarta.servlet-api</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        <dependency>
            <groupId>com.github.paulcwarren</groupId>
            <artifactId>spring-content-rest-boot-starter</artifactId>
            <version>3.0.1</version>
            <exclusions>
                <exclusion>
                    <groupId>commons-logging</groupId>
                    <artifactId>commons-logging</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>jakarta.servlet</groupId>
                    <artifactId>jakarta.servlet-api</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
<!--
        <dependency>
            <groupId>com.github.paulcwarren</groupId>
            <artifactId>spring-content-fs-boot-starter</artifactId>
            <version>3.0.1</version>
        </dependency>
-->

        <!-- Log -->
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-log4j2</artifactId>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.dataformat</groupId>
            <artifactId>jackson-dataformat-yaml</artifactId>
            <version>2.14.2</version>
        </dependency>

        <!-- Tooling -->
        <dependency>
            <groupId>org.springframework.data</groupId>
            <artifactId>spring-data-rest-hal-explorer</artifactId>
        </dependency>

        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-devtools</artifactId>
            <scope>runtime</scope>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-configuration-processor</artifactId>
            <optional>true</optional>
        </dependency>

        <!-- Test -->
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-test</artifactId>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.springframework.security</groupId>
            <artifactId>spring-security-test</artifactId>
            <scope>test</scope>
        </dependency>

    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <configuration>
                    <source>16</source>
                    <target>16</target>
                </configuration>
            </plugin>

            <plugin>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-maven-plugin</artifactId>
                <executions>
                    <execution>
                        <goals>
                            <goal>repackage</goal>
                        </goals>
                    </execution>
                </executions>
                <configuration>
                    <imageName>${project.groupId}/${project.artifactId}:${project.version}</imageName>
                </configuration>
            </plugin>

        </plugins>
    </build>

    
</project>
```

**Comment by paulcwarren** [comment]
OK.  I added a [sample project](https://github.com/paulcwarren/spring-content-issue-1374) that shows the method level preauthorize do get called.  That said, I do agree there is some strange behavior depending on the how the class-level pre-authorize is specified.

Should just need to git clone and run the GettingStartedTest.  Modify as needed to test the different scenarios.

**Comment by Nick-Wunderdog** [comment]
I have little new information. On the repo of the content parent. 

```java
@RepositoryRestResource
@PreAuthorize("false")
public interface AttachmentRepository extends CrudRepository<Attachment, Long> {
    @Override
    @PreAuthorize("isFullyAuthenticated()")
    @PostAuthorize(HAS_ROLE_READ + " OR returnObject.orElse(new Attachment()).parent.parent.owners.contains(authentication.name) OR authentication.name.equals(returnObject.orElse(new Attachment()).parent.creator)")
    Optional<Attachment> findById(Long id);
```

Which works perfectly via the Spring Data REST API to get the Attachment entity : 
`GET /api/programs/1` 

But calling 
`POST /api/programs/1/content` 
 (omitting boilerplate) 

Server return HTTP 500 with stacktrace : 

023-05-03 18:52:27,254 DEBUG o.s.s.a.m.AuthorizationManagerAfterMethodInterceptor [http-nio-8081-exec-1] Authorizing method invocation ReflectiveMethodInvocation: public abstract java.util.Optional xxx.repository.AttachmentRepository.findById(java.lang.Long); target is of class [jdk.proxy4.$Proxy208]
2023-05-03 18:52:27,263 ERROR o.a.j.l.DirectJDKLog [http-nio-8081-exec-1] Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed: java.lang.NullPointerException: Cannot invoke "java.util.Optional.orElseThrow(java.util.function.Supplier)" because "domainObj" is null] with root cause
java.lang.NullPointerException: Cannot invoke "java.util.Optional.orElseThrow(java.util.function.Supplier)" because "domainObj" is null
	at internal.org.springframework.content.rest.controllers.resolvers.DefaultEntityResolver.findOne(DefaultEntityResolver.java:187) ~[spring-content-rest-3.0.1.jar:?]
	at internal.org.springframework.content.rest.controllers.resolvers.DefaultEntityResolver.resolve(DefaultEntityResolver.java:107) ~[spring-content-rest-3.0.1.jar:?]
	at internal.org.springframework.content.rest.controllers.resolvers.EntityResolvers.resolve(EntityResolvers.java:38) ~[spring-content-rest-3.0.1.jar:?]
	at internal.org.springframework.content.rest.controllers.ResourceHandlerMethodArgumentResolver.resolveArgument(ResourceHandlerMethodArgumentResolver.java:111) ~[spring-content-rest-3.0.1.jar:?]
	at org.springframework.web.method.support.HandlerMethodArgumentResolverComposite.resolveArgument(HandlerMethodArgumentResolverComposite.java:122) ~[spring-web-6.0.8.jar:6.0.8]

**Comment by Nick-Wunderdog** [comment]
It seems that when ContentStore code calls it's parent Entity repository ".findById()" something is different and missing compared to calling it via the Spring Data REST path.

if the Entity repo `findById()` has 

`@PostAuthorize(returnObj.getOrElse(new File()).nestedObject.nestedObject.collectionField.contains("xxxx"))`
Which will work 100% in all use cases when used via Spring Data REST endpoint. But when used via Content Store there is NPE thrown and actually AccessDenied - exception is on top of the stack trace. But for some reason this is not delegated as AccessDenied - exception to POST (/files/1/content) HTTP request, but an 500 instead saying "returnObj is null in "returnObj.orElse() " 
Which is very confusing as  returnObj is Optional and can not be null, it can be empty, but not null

**Comment by Nick-Wunderdog** [comment]
I added an PR:  
https://github.com/paulcwarren/spring-content/pull/1400

Logging and catching ACE are optional, I guess it should work just with initializing the optional as empty.

**Comment by Nick-Wunderdog** [comment]
I guess the RP did not make to 3.0.2 ?

**Comment by paulcwarren** [comment]
It did not but it sounds like it resolves your issue?  If so I can cut another release for you.

**Comment by Nick-Wunderdog** [comment]
I have not tried 3.0.2 yet. It was released day before my PR was merged.
That is why I asked.

**Comment by paulcwarren** [comment]
The fix is not in 3.0.2 so that wont work.  I was asking if the PR represents a completed fix for you?  If so I will cut a 3.0.3 release for you.

**Comment by paulcwarren** [comment]
Hi @Nick-Wunderdog , I cut 3.0.3 release for you yesterday.  I haven't created the github release yet but its out there on maven central.

**Comment by paulcwarren** [comment]
Fixed in 3.0.3
